### ISSUE TITLE
Show speed limit

### ISSUE THREAD

**Issue by ccptr** [organicmaps/organicmaps#952]
If the speed limit is known it'd be nice to be able to see it.

**Comment by biodranik** [comment]
Even better would be an easy and convenient way to add/improve speed limits along the way :)

**Comment by Orejane** [comment]
EStimation of Arrival Time should be based on speed limit

**Comment by biodranik** [comment]
Not every road in OSM has existing speed limit tag set. During the ride you may see the (probably outdated) speed limit, then see nothing. And then see a speed limit again.

**Comment by ccptr** [comment]
> EStimation of Arrival Time should be based on speed limit

Is it not?

**Comment by Orejane** [comment]
> > EStimation of Arrival Time should be based on speed limit
> 
> Is it not?

 as speed limit does not show I cannot tell it for sure, but  my impression is that estimation is wrong in town

**Comment by Nic787** [comment]
> > > EStimation of Arrival Time should be based on speed limit
> > 
> > 
> > Is it not?
> 
> as speed limit does not show I cannot tell it for sure, but my impression is that estimation is wrong in town

Remind me of a time in Iceland where the estimated time of a trip was 1h30, but Google Maps was giving me 2h30... Guess who was right...

**Comment by ccptr** [comment]
>I'm often surprised by the 20-minute estimates when the actual ride to the town takes over one and half hours.

Are you sure OSM has speed limit data for those roads? There were a couple highways in between two towns that I often routed between that didn't have `maxspeed` set, and as a result OsmAnd took me on some weird routes with completely inaccurate ETAs. That sounds like it's assuming 100 km/h on 30-50 km/h roads or something.

**Comment by Talkless** [comment]
> Not every road in OSM has existing speed limit tag set

Maybe it would be possible to incorporate implicit default database: https://wiki.openstreetmap.org/wiki/Default_speed_limits

Biggest question if, does map processor "knows" if road is urban (within place area) or rural using OSM data?

**Comment by AntonM030481** [comment]
You can here OSM map here, where each road is marked with it's speed limit (if info is present):
https://www.openstreetbrowser.org/#map=17/50.07620/19.89539&categories=car_maxspeed
You can easily add speed limit there if no info is present.
![image](https://user-images.githubusercontent.com/85622715/166874623-7d33c0cf-ad41-47e4-8927-0c8f3d5658fe.png)

**Comment by AntonM030481** [comment]
@biodranik so, why not to show the speed limit from OSM data?

**Comment by MisterY** [comment]
Perhaps the title should be adjusted to narrow down to showing the speed limit in navigation mode? It would really help driving in the city. Especially now that the routing feel better and the routes are different (and seem better) to other apps that I have, but which also use OSM data.

**Comment by biodranik** [comment]
@AntonM030481 implement it ) it was partially done already.

**Comment by bitboy85** [comment]
If a speed limit is known, it should be shown :)
The data might be outdated, but this argument is also true for other parts of the map:
Closed roads, exits, new lanes, new crossings...
If someone notice wrong data, the person is at least able to correct it. (Or making a note on osm)

**Comment by AntonM030481** [comment]
@biodranik challenge accepted 😄

**Comment by biodranik** [comment]
> If a speed limit is known, it should be shown :)
> 
> The data might be outdated, but this argument is also true for other parts of the map:
> 
> Closed roads, exits, new lanes, new crossings...
> 
> If someone notice wrong data, the person is at least able to correct it. (Or making a note on osm)

The problem is what to show when the speed limit is not known. An empty circle? A guess?

**Comment by MisterY** [comment]
> The problem is what to show when the speed limit is not known. An empty circle? A guess?

Nothing?

**Comment by AntonM030481** [comment]
Waze does it like this:
![image](https://user-images.githubusercontent.com/85622715/172839670-de557d93-2aa9-4476-8b8b-13361685aa50.jpeg)
If speed limit is unknown - it's not shown.

**Comment by Talkless** [comment]
> The problem is what to show when the speed limit is not known. An empty circle? A guess?

Maybe a transparent, gayish circle with "?" inside of it, but maybe better just don't show it.

**Comment by tomicdusan** [comment]
Yeah, this is really good to be displayed on screen if you use navi abroad (traveling, etc), pretty please? :)

**Comment by pm4rcin** [comment]
@AntonM030481 what stops you from implementing it? Is it lack of free time or you still haven't decided on how it should look (style, size etc.)? If the latter here's my opinion with rationale:
> I think that something similiar to Magic Earth's implementation would be best (maybe the size and/or style could be adjusted). Rationale:
When you want to check the speed limit you usually want to see it as fast as possible and focus on driving after seeing it. So it should be the shortest path for eyes. From my experience it's faster to focus on something that's on top because you generally have phone below line eyes and the path to look on right or right with slight down is fast and you don't have to move the eyes to the down corner of your phone (there's probably a paper on that but who cares :D).

**Comment by alensiljak** [comment]
Is there a way to use even the text version, i.e. in a format [current speed/speed limit] or similar? The need for seeing this information is relatively high, irrespective of the presentation of it. A visual speed-limit sign would be nice but having the information on the dashboard is a priority. If the feature is waiting for nice graphics I'd be all up for having whatever is available right now, and work on the visuals later.

**Comment by biodranik** [comment]
You can check https://github.com/organicmaps/organicmaps/pull/2963 and implement the same for Android.

**Comment by alensiljak** [comment]
> You can check #2963 and implement the same for Android.

Not so simple, unfortunately. It seems that the ObjectC code for iPhone has direct access to the FollowingInfo class, whereas in Android it goes through JNI.

There is a problem when I'm, in `Framework.cpp`, method `Java_app_organicmaps_Framework_nativeGetRouteFollowingInfo` trying to pass `info.m_speedLimitMps` to the constructor for `RoutingInfo`. Always getting an Abort in native code so I don't see in the debugger what's the problem is. I've added the parameter to the RoutingInfo constructor, naturally.

**Comment by biodranik** [comment]
On Android, the reason for the error/abort is always printed in the log. You can also setup JNI debugger in Android Studio:
https://developer.android.com/studio/debug

**Comment by alensiljak** [comment]
Thanks for the tip! I did not really understand the obscure syntax for GetConstructorID, it seems.

Continuing in #3817.

**Comment by alensiljak** [comment]
Now that the #3817 is more-or-less in place, I see two suggestions:

1. Add a visual indicator for the speed limit. Eventually a setting that controls the displays of it (on/off).
2. Indicate overstepping the limit based on a calculation (including a threshold) and indicate with an audio and visual signal.

These can be created and discussed as separate issues, I believe.

**Comment by bitboy85** [comment]
Additional suggestions:
- show speed limit in non-navigation mode
- unknown speed limits might be represented by a red circle with a question mark or even better: use the countrys default limit for this type of road because many roads in citys or towns are not tagged
- note that some highways in germany (and maybe other countries) have no limit at all.

**Comment by Talkless** [comment]
> unknown speed limits might be represented by a red circle with a question mark or even better: use the countrys default limit for this type of road because many roads in citys or towns are not tagged

Not sure if it's easy to distinguish if your'e in city or not?

Maybe show table of speed limits, like ones you see when crossing boarder, showing speed limits for cities, higways, motorways, etc..? Though that might be not that easy to specify too...?

**Comment by bitboy85** [comment]
I think its perfectly doable. Osm usually uses administrative bounderies for countries, states, cities... so you need to do 3 things: 1. check if your position is within such an area. 2. During driving check if you cross an administrative boundary (to update rules). Set the information in relation to the road type. For example highways "within" might have no limit.
The needed algo might be already there.

**Comment by pm4rcin** [comment]
> I think its perfectly doable. Osm usually uses administrative bounderies for countries, states, cities... so you need to do 3 things: 1. check if your position is within such an area. 2. During driving check if you cross an administrative boundary (to update rules). 3. Set the information in relation to the road type. For example highways "within" might have no limit. The needed algo might be already there.

Default limit is not a good idea because it's not like city boundary is the same as the Urban speed limit boundary. In Poland we have green signs for city boundary but the white sings aka Urban area are separate so you can be in the city boundary but outside Urban area (it usually means there aren't homes and pedestrian ways nearby (simplified definition)). Better idea is to add speed limits to missing roads. To view them check [Anton's link](https://github.com/organicmaps/organicmaps/issues/952#issuecomment-1118227251).

**Comment by Talkless** [comment]
> In Poland we have green signs for city boundary but the white sings aka Urban area are separate

Good point. In Lithuania, blue signs means there's no city limit while crossing urban area, white signs means you go at 50.

**Comment by biodranik** [comment]
At the moment, only CarPlay has implemented the speed limit properly:
![image](https://user-images.githubusercontent.com/170263/221775717-264b88eb-00f3-474c-b1c1-0de59564bc11.png)

iPhone shows the speed limit in the bottom left corner. Android doesn't show it at all.

It would be great to use the same approach as is used in CarPlay everywhere.

**Comment by biodranik** [comment]
CarPlay and Android Auto display speed limit properly (@AndrewShkrob right?)

iOS displays the speed limit near the current speed, at the left bottom of the screen (in LTR mode), and changes its color to red if the speed limit exceeds the allowed one.

At least the same "minimal" implementation for Android is very desirable until we find resources to redesign the UI properly. Any volunteers?

**Comment by AndrewShkrob** [comment]
Android Auto doesn't support speed limit. We need to render it manually.

**Comment by biodranik** [comment]
@AndrewShkrob is there a separate issue for speed limit in AA?

**Comment by oleg-PK** [comment]
Related to #6560
